byway/src/main/webapp/destinations.js

```diff
@@ -0,0 +1,201 @@
+      }
+      const icon = {
+        url: place.icon,
+        size: new google.maps.Size(71, 71),
+        origin: new google.maps.Point(0, 0),
+        anchor: new google.maps.Point(17, 34),
+        scaledSize: new google.maps.Size(25, 25)
+      };
+      // Create a marker for each place.
+      markers.push(
+        new google.maps.Marker({
+          map,
+          icon,
+          title: place.name,
+          position: place.geometry.location
+        })
+      );
+
+      if (place.geometry.viewport) {
+        // Only geocodes have viewport.
+        bounds.union(place.geometry.viewport);
+      } else {
+        bounds.extend(place.geometry.location);
+      }
+    });
+    map.fitBounds(bounds);
+}
+
+/*
+* fetches start location and destinations from DestinationsServlet and adds to DOM
+*/
+function getLocations(){
+  fetchDestinations().then((userLocations) => {
+    document.getElementById('start-location').innerText = "Start Location :" + userLocations.start;
+    const container = document.getElementById('destinations-container');
+    container.innerText = "Destinations:";
+    let destinationArray= userLocations.destinations;
+    destinationArray.forEach((destination) => {
+      let destinationToAdd = document.createElement('p');
+      destinationToAdd.innerText = destination;
+      container.appendChild(destinationToAdd);
+    }) 
+  });
+}
+
+/* 
+* fills Start location Searchbox with previously input
+*/
+function getStartDestination(){
+  fetchDestinations().then((userLocations)=>{
+    if (userLocations.start == null){
+      document.getElementById('start-search-box').value = "";
+    }
+    else{
+      document.getElementById('start-search-box').value = userLocations.start;
+    }
+  });
+}
+
+/* 
+* fetches data from servlet
+*/
+function fetchDestinations(){
+  return fetch('/api/destinations').then(response => response.json());
+}
+
+/* 
+* Gets users current location 
+*/
+function getCurrentAddress(){
+  const geocoder = new google.maps.Geocoder();
+  geocoder.geocode({ 'location': userlatlng}, (results, status) => {
+    if (status === "OK") {
+      if (results[0]) {
+          console.log(results[0].formatted_address);
+        document.getElementById("start-search-box").value=results[0].formatted_address;
+      } else {
+          window.alert("No results found");
+        }
+    } else {
+        window.alert("Geocoder failed due to: " + status);
+      }
+  });
+}
+
+/* 
+* add event listener for submit button
+*/
+window.onload = function(){
+  document.getElementById('user-input-form').addEventListener('submit', (event) => {
+    event.preventDefault();
+    const params = new URLSearchParams();
+    params.append('start-location', document.getElementById('start-search-box').value); 
+    params.append('destinations', document.getElementById('destinations-search-box').value);  
+    fetch('/api/destinations', {method: 'POST', body:params}).then(function(response){
+      return response.text();
```

Comment:
If you return json from the post action, you could change `getLocations` and`getStartDestination` to `updateLocations(locationsData)` and `updateStartDestination(locationData)` and use them like:

```js
fetch(<...>).then(response => {
  updateLocations(response.json());
  updateStartDestination(response.json());
});
```

Notice that this `update` rather than `get` setup (where the function to update the page takes the data to use as an argument rather than fetching it itself) makes it possible to also avoid multiple `fetch` calls when setting up the page, since you can have:

```js
function initDestinationsPage() {
  fetchDestinations().then(destinations => {
    updateLocations(destinations);
    updateStartDestination(destinations);
  });
}
```

Note that in JS it's possible to take a variable number of arguments in a function, because any arguments that aren't provided will have the special value `undefined`, so it's possible to have `fetchDestinations` do both get and post:

```js
function fetchDestnations(optPostData) {
  let destinationsPromise;
  if (optPostData) {
    destinationsPromise = fetch('/api/destinations', {method: 'POST', body: optPostData});
  } else {
    destinationsPromise = fetch('/api/destinations');
  }
  return destnationsPromise.then(response => response.json());
}

Reply:
done